### PR TITLE
run: Support endpoint enrichment

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
This commit introduces a new endpoint type that can be used by the
events generated by the gadgets. The gadgets need to fill in the IP
address (IPv4 or IPv6), the version (4 or 6) and the port (in host byte
order) for each endpoint. Then the user space part automatically enriches
that with the KubeIPResolver operator.

One virtual column is added per endpoint, and it's enriched and
formatted according to the data available:

```bash
$ ./kubectl-gadget run --prog @./gadgets/trace_tcpconnect_x86.bpf.o --definition @./gadgets/trace_tcpconnect.yaml
INFO[0000] Experimental features enabled
NODE                   NAMESPACE             POD                   CONTAINER             PID     TASK        SRC                      DST
ubuntu-hirsute         default               mypod6                mypod6                182408  wget        p/default/mypod6:44632   r/1.1.1.1:80
ubuntu-hirsute         default               mypod6                mypod6                182408  wget        p/default/mypod6:33464   r/1.1.1.1:443
ubuntu-hirsute         default               mypod6                mypod6                182444  wget        p/default/mypod6:37184   s/default/nginx:80
```

Custom columns are available for different fields inside the endpoint (addr, port, etc):

```bash
$ ./kubectl-gadget run --prog @./gadgets/trace_tcpconnect_x86.bpf.o --definition @./gadgets/trace_tcpconnect.yaml -o columns=node,pod,src.addr,src.port,dst.addr,dst.port
INFO[0000] Experimental features enabled
NODE                                   POD                                    SRC.ADDR             SRC.PORT            DST.ADDR            DST.PORT
ubuntu-hirsute                         mypod6                                 172.16.118.129       51828               1.1.1.1             80
ubuntu-hirsute                         mypod6                                 172.16.118.129       45184               1.1.1.1             443
```

Endpoints are also presented in the json/yaml/jsonpretty output modes:

```bash
$ ./kubectl-gadget run --prog @./gadgets/trace_tcpconnect_x86.bpf.o --definition @./gadgets/trace_tcpconnect.yaml -o json | jq
...
    {
      "src": {
        "port": 38548,
        "v": 4,
        "namespace": "default",
        "kind": "pod",
        "name": "mypod6",
        "addr": "172.16.118.191",
        "proto": "TCP"
      },
      "dst": {
        "kind": "svc",
        "namespace": "default",
        "v": 4,
        "addr": "10.108.196.40",
        "name": "nginx",
        "port": 80,
        "proto": "TCP"
      },
      "runtime": {
        "runtimeName": "containerd",
        "containerId": "23d304dfed76c670696cb532f5f6efed7cda9e91e7884be35ec227ef5f613d80",
        "containerName": "",
        "containerImageName": "docker.io/library/busybox:latest"
      },
      "node": "ubuntu-hirsute",
      "namespace": "default",
      "pod": "mypod6",
      "container": "mypod6",
      "hostnetwork": false,
      "timestamp": "1970-01-01T00:00:00.000000000Z",
      "mntns": 4026532782,
      "pid": 11116,
      "task": "wget",
      "uid": 0,
      "gid": 0
    }
```

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/1917

## TODO
- [x] Make extractor generic not to marshal integers as strings in json (https://github.com/inspektor-gadget/inspektor-gadget/pull/1817#pullrequestreview-1530468503)
- [ ] Avoid running operators all the time (https://github.com/inspektor-gadget/inspektor-gadget/pull/1825#discussion_r1259537234) (Future PR?)

### CODE REVIEW

File: pkg/gadgets/run/types/types.go
Comment (burak-ok) [inline pkg/gadgets/run/types/types.go:24]: Other PR: Move the `Version` field into the `L3Endpoint`?

File: pkg/gadgets/common/types.h
Comment (mauriciovasquezbernal) [reply pkg/gadgets/common/types.h]: Done. ("Funny" thing: It took me almost 1 day to get it right after hitting every single possible problem).

File: pkg/gadgets/run/types/types.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/types.go]: I did in this one as part of the reworking to have types for L3 and L4 endpoints in C.

File: pkg/gadgets/run/types/types.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/types.go:45]: Good point. We have the same problem for operators checking `SetMountNsMap`. I think we would need to change the initialization logic to have a sort of "PreRun" to allow the gadget to configure itself, then the operators should not check the gadget desc but the tracer instance to see if they can operate on that. (Not only based on the implementation of an interface but on the return value of a method).  @flyth any thoughts on this?

Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/types.go:45]: I totally agree with the first statement: users should be able to disable expensive operators. On the second one, I think `CanOperateOn` still needs some knowledge about the gadget: even if the operator is enabled, we should avoid creating and instance of it if it's not compatible with a given gadget.

Comment (flyth) [reply pkg/gadgets/run/types/types.go:45]: An operators' Instantiate() can return nil in case it doesn't want to be enabled (at least that's how it was planned). In Instantiate() you should have all the necessary information (e.g. operator params), right? CanOperateOn should be just used for the compatibility check.

File: pkg/gadgets/common/types.h
Comment (mauriciovasquezbernal) [reply pkg/gadgets/common/types.h:16]: If I remove the padding here and in `type l3EndpointT struct` it doesn't work (It seems to me that golang and C use different padding). IMO it's good to be sure both structures have the exact same memory layout by manually adding padding. 

Another option was to use cgo but I'm not totally sure it fixes the problem:
- The structure will be compiled by two different compilers: clang (ebpf) and gcc (user space), then I don't think it's guaranteed that both have the exact same layout. 
- I'm not sure it's worth it to introduce cgo just to get the definition of a structure 

Thoughts on that?

File: pkg/gadgets/common/types.h
Comment (mauriciovasquezbernal) [reply pkg/gadgets/common/types.h:24]: I added a proto field. It's hidden by default but I agree it can be useful for gadgets that support multiple protocols.

**Comment by mauriciovasquezbernal** [comment]
This is ready for review folks, thanks!

File: pkg/gadgets/run/tracer/tracer.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/tracer.go]: Right. Fixed.

File: pkg/gadgets/run/tracer/tracer.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/tracer.go]: I used a similar approach. Thanks for the suggestion.

File: pkg/gadgets/run/tracer/tracer.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/tracer.go]: Right. Fixed.

File: pkg/types/types.go
Comment (mauriciovasquezbernal) [reply pkg/types/types.go:231]: Added a ref to the iana document defining them.

File: pkg/types/types.go
Comment (mauriciovasquezbernal) [reply pkg/types/types.go:248]: I'm not sure. It seems a bit weird to print the proto there. I'd rather avoid it for now.

**Comment by mauriciovasquezbernal** [comment]
> But I find the code to get the L3/L4 endpoint a bit too complicated.

Unfortunately I haven't find a better way to handle it. Suggestions are welcome.

File: pkg/types/types.go
Comment (alban) [inline pkg/types/types.go:235]: #1943 introduces the same function ProtoString(). If #1943 is merged first, this should be refactored to avoid duplicate functions.

Comment (mauriciovasquezbernal) [reply pkg/types/types.go]: Fixed. Thanks for the heads up.

**Comment by mauriciovasquezbernal** [comment]
This is ready for reviewing folks. It's still missing to fix the issue about having operators always run (https://github.com/inspektor-gadget/inspektor-gadget/pull/1825#discussion_r1259537234), but that requires some more work and I'd prefer to tackle that in another PR.

File: pkg/gadgets/run/types/types.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/types.go:45]: I did some experiments in https://github.com/inspektor-gadget/inspektor-gadget/commit/4a4f3db719b68a98f01efa18f701de10cacca972. It doesn't look like a perfect solution to me but something in that direction could work fine. My main concern is about the overlap with `CanOperate()`, should we keep both of them?

File: cmd/common/image/list.go
Comment (mauriciovasquezbernal) [reply cmd/common/image/list.go:97]: I did that manually. I'll consider using `gopatch` next time.

**Comment by mauriciovasquezbernal** [comment]
> I still find the code heavy but I still do not have another solution, so I think we are tied with it.

I share the same concerns. This is still in a very early development phase, so I don't want to expend too much time trying to write perfect code, I'd rather prefer some code that works and then iterate to make it better. 

> Do we have some tests for containerized gadgets? I would like to avoid the case where we translate plenty of gadgets to this, deprecate the old and just a whole red CI.

We have an integration tests for trace_open (https://github.com/inspektor-gadget/inspektor-gadget/blob/main/integration/inspektor-gadget/run_trace_open_test.go), and we'll be adding more tests as we convert the gadgets. 

Thanks for the review!

File: pkg/gadgets/run/types/types.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/types/types.go:45]: The issue is that they work at different levels. `CanOperateOn` is more static: it only uses interfaces and data types to understand if a given operator can work on a gadget. It made sense for "built-in" gadgets, but the `run` gadget is do dynamic that depending on the actual gadget being run some operator make sense and others don't. 

I'd say that we can keep both of them right now, once we start deprecating (and removing) built-in gadgets the need for  `CanOperateOn`  could disappear.

**Review by alban — APPROVED** [review]
Thanks!
I added a few things to change, but LGTM after that.

I am not so sure about the changes in the columns pkg, but I don't want to block this PR waiting for another review.

File: pkg/gadgets/run/tracer/tracer.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/tracer.go:237]: Not for now, it's an internal detail.

File: pkg/gadgets/run/tracer/run.go
Comment (mauriciovasquezbernal) [reply pkg/gadgets/run/tracer/run.go:255]: Good catch, thanks!

**Pull request merged**
